test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,329 @@
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch is interspersed among the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 500),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch contains some of already-existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 500),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch with both DM and stream messages -> ignores DM, processes stream messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+        ];
+        setupModel(existingMessages);
+
+        final dmMessage = eg.dmMessage(from: eg.otherUser, to: [eg.selfUser], id: 200);
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          dmMessage,
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages..remove(dmMessage)]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+    });
+
+    group('multiple trackers', () {
+      test('batch goes before the existing messages', () {
```

Comment:
I don't think it's all that helpful to take the single-tracker tests and make a copy of each of them that distributes across a couple of different topics. What I asked for at https://github.com/zulip/zulip-flutter/pull/692#discussion_r1683760587 was:
> there should also be a couple of test cases that show that `handleMessages` correctly handles taking messages that belong in several different trackers, and distributing them across those.

These test cases don't really do that. They don't test that `handleMessages` correctly handles:
* having messages in several streams
* or messages from several senders
* or creating a new tracker where there wasn't one already;

really the one thing they add is testing that it handles getting messages in multiple topics of one stream.

Conversely there isn't a need to repeat the testing of all the different ways the old and new messages in a given tracker can be related; that logic all operates within each tracker separately, so the single-tracker tests cover it.

These tests do add a lot of lines of code. One effect of that is that it takes some work to scan through and see all the things the tests are saying, in order to get the picture of what scenarios are covered and what scenarios aren't. One reason making tests concise and focused (as discussed in comments just above this one) is important is that it helps with detecting missing scenarios, like here.

I'll push another commit replacing these with tests that cover those.